### PR TITLE
Limit request size per peer

### PR DESCRIPTION
**Pull request by asdacap**
Fixes | Closes | Resolves #

> Anything marked as optional that you didn't need to fill in, please remove it from the PR description. Choose one of the keywords above to refer to the issue this PR solves, followed by the issue number (e.g Fixes # 666). If there is no issue, remove the line. Remove this note after reading.

## Changes:
-

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [ ] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [ ] No

**Comments about testing , should you have some** (optional)

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/Nethermind/Nethermind.Core/Collections/IReadOnlyListExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+namespace Nethermind.Core.Collections;
+
+public static class IReadOnlyListExtensions
+{
+    public static IReadOnlyList<T> CappedTo<T>(this IReadOnlyList<T> readOnlyList, int length)
+    {
+        if (length > readOnlyList.Count)
+        {
+            return readOnlyList;
+        }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Collections/IReadOnlyListExtensions.cs:12]: The original type like `List<>` can still grow, so we shouldn't ever use this path

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V63/Eth63ProtocolHandler.cs [hunk 2]
```diff
@@ -142,9 +151,40 @@ public override async Task<TxReceipt[][]> GetReceipts(IReadOnlyList<Keccak> bloc
...
+            }
+            catch (Exception)
+            {
+                _getReceiptsCurrentBatchSize = Math.Max(
+                    (int)Math.Floor(startingReceiptsCountLimit / GetReceiptsBatchSizeAdjustmentFactor),
+                    GetReceiptsMinBatchSize
+                );
+                throw;
+            }
         }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V63/Eth63ProtocolHandler.cs:188]: We have this code very similar in two places, should we make it reusable somehow?

**Review by LukaszRozmej — COMMENTED** [review]
Are we progressing with this?

File: src/Nethermind/Nethermind.Core/Collections/IReadOnlyListExtensions.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Collections/IReadOnlyListExtensions.cs:12]: Why? Its a read only list?

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V63/Eth63ProtocolHandler.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V63/Eth63ProtocolHandler.cs:188]: Probably

**Comment by asdacap** [comment]
> Are we progressing with this?

I thought I should push a PR for sync metrics first to see the reduced timeout.

**Pull request closed without merging**